File: tests/golden_outputs/design_update_features.c
#include <stdio.h>
void test_casts(void) {
    int x = 42;
    float y = ((float)(x));
    void *p = ((void*)(x));
    char c = ((char)(65));
}
void test_escaped_identifiers(void) {
    int _ = 5;
    char *FILE__ = "test.c";
    int x = 10;
}
#define MAX 100
#define MIN 0
#undef MAX
#undef MIN
#define DEBUG 1
#define VERBOSE 2
#undef DEBUG
#undef VERBOSE
_Atomic int counter;
_Atomic int *ptr;
_Alignas(16) int aligned_var;
_Alignas(64) int aligned_array[10];
void test_alignof(void) {
    size_t a = _Alignof(struct Node);
    size_t b = _Alignof(int);
}
struct Node {
    int data;
    struct Node *next;
};
void test_static_assert(void) {
    _Static_assert(sizeof(struct Node) > 0, "Node size must be positive");
}
_Thread_local int tls_var;
static _Thread_local int static_tls;
struct Outer {
    int a;
    struct {
        int x;
        int y;
    };
    int b;
};
struct Tagged {
    int tag;
    union {
        int i;
        float f;
        double d;
    };
};
struct Widget {
    enum {
        RED = 0,
        GREEN = 1,
        BLUE = 2,
    } color;
    int value;
};
enum {
    OK = 0,
    ERROR = 1,
    PENDING = 2,
} global_status;
typedef int BinaryFunc(int, int);
typedef void UnaryFunc(double);
typedef int NoArgFunc(void);
int add(int, int);
int (*callback)(int, int);
void (*handler)(void);
typedef int (*BinOp)(int, int);
int apply(int (*f)(int, int), int a, int b) {
    return f(a, b);
}
void test_function_cast(void) {
    void *ptr = ((void*)(0));
    int (*fp)(int, int) = ((int (*)(int, int))(ptr));
}
int main(void) {
    return 0;
}

File: tests/fixtures/design_update_features.py
from stdio import *

# ============================================================================
# Test 1: Casts - cast[T](expr)
# ============================================================================

def test_casts() -> void:
    x: int = 42
    y: float = cast[float](x)
    p: -void = cast[-void](x)
    c: char = cast[char](65)

# ============================================================================
# Test 2: Escaped Identifiers - __ID → ID
# ============================================================================

def test_escaped_identifiers() -> void:
    # ___ → _ in C
    ___: int = 5

    # __FILE__ → FILE__ in C
    __FILE__: -char = "test.c"

    # Regular identifiers work as before
    x: int = 10

# ============================================================================
# Test 3: #undef Support - del NAME
# ============================================================================

MAX: macro = 100
MIN: macro = 0

del MAX
del MIN

# Multiple undefs
DEBUG: macro = 1
VERBOSE: macro = 2
del DEBUG, VERBOSE

# ============================================================================
# Test 4: C11 Qualifiers
# ============================================================================

# 4.1: atomic[T]
counter: atomic[int]
ptr: -atomic[int]

# 4.2: alignas[N, T]
aligned_var: alignas[16, int]
aligned_array: alignas[64, list[int, 10]]

# 4.3: alignof[T]
def test_alignof() -> void:
    a: size_t = alignof[type[Node]]
    b: size_t = alignof[int]

# 4.4: static_assert
class Node:
    data: int
    next: -type[Node]

def test_static_assert() -> void:
    static_assert(sizeof(type[Node]) > 0, "Node size must be positive")

# 4.5: thread_local[T]
tls_var: thread_local[int]
static_tls: static[thread_local[int]]

# ============================================================================
# Test 5: Anonymous Embedded Aggregates
# ============================================================================

class Outer:
    a: int

    # Anonymous embedded struct
    class _:
        x: int
        y: int

    b: int

class Tagged:
    tag: int

    # Anonymous embedded union
    class _(Union):
        i: int
        f: float
        d: double

# Anonymous enum in struct
class Widget:
    @var(color)
    class _(Enum):
        RED = 0
        GREEN = 1
        BLUE = 2

    value: int

# Top-level anonymous enum
@var(global_status)
class _(Enum):
    OK = 0
    ERROR = 1
    PENDING = 2

# ============================================================================
# Test 6: Function and Function Pointer Types
# ============================================================================

# 6.1: Function type - (args)(result)
type BinaryFunc = (int, int)(int)
type UnaryFunc = (double,)(void)  # Single-element tuple needs trailing comma
type NoArgFunc = ()(int)

# Function declaration with new syntax
add: (int, int)(int)

# 6.2: Function pointer type - -(args)(result)
callback: -(int, int)(int)
handler: -()(void)  # Function pointer: no args -> void

# Type alias for function pointer
type BinOp = -(int, int)(int)

# Function taking function pointer as argument
def apply(f: -(int, int)(int), a: int, b: int) -> int:
    return f(a, b)

# ============================================================================
# Test 7: Combined Examples
# ============================================================================

# Cast to function pointer
def test_function_cast() -> void:
    ptr: -void = cast[-void](0)
    fp: -(int, int)(int) = cast[-(int, int)(int)](ptr)

def main() -> int:
    return 0
